src/cc/frontends/clang/b_frontend_action.cc

```diff
@@ -911,7 +911,11 @@ void BTypeConsumer::HandleTranslationUnit(ASTContext &Context) {
       if (fe_.is_rewritable_ext_func(F)) {
         for (auto arg : F->parameters()) {
           if (arg == F->getParamDecl(0)) {
-            probe_visitor1_.set_ctx(arg);
+            string type = arg->getType().getAsString();
+            if (type == "struct pt_regs *" ||
+                type == "struct bpf_raw_tracepoint_args *" ||
+                type.substr(0, 19) == "struct tracepoint__")
```

Comment:
This seems a bit fragile. Is there a better way to check the type of the context argument?

Reply:
This indeed fragile. The best way is to know the attachment of the external function so based on what kind of attachment it is (`raw_tracepoint`, `tracepoint`, `kuprobe`, or others), we will be able to know whether the first parameter is a ctx or not. Unfortunately, we did not have such information in program load time.

So I think this change sounds good. It may miss some cases if user have `void *ctx` as the argument and later on cast `ctx` to something else, but it won't rewrite incorrectly.

Reply:
> It may miss some cases if user have void *ctx as the argument and later on cast ctx to something else, but it won't rewrite incorrectly.

I was tempted to blacklist instead of whitelisting for a moment (i.e., do not add ctx pointer if it points to a packet), but that seemed even more fragile, especially when new types of context will be added...